File: src/skillhub/manifest.py
"""Deployments manifest — manifests/deployments.json 读写 + 删除链。

schema v1:
{
  "version": 1,
  "records": [
    {
      "skill": "canvas-design",
      "machine": "mac-main",
      "agent": "TeleAgent",
      "deploy_path": "/Users/ss/.config/TeleAgent/skills/canvas-design",  # 绝对路径(已按 machine home 展开)
      "deployed_at": "2026-08-15T02:00:00Z",
      "method": "cp",                       # cp | ln
      "ir_hash": "sha256:...",              # body hash(零损耗跨机验证)
      "note": "description truncated ...",  # emitter note(可空)
      "pending_deletion": false             # 跨机删除标记: true = 等那台机器下次 sync 时删
    }
  ]
}

删除链(决策 5/6):
1. skillhub rm <name> / skill 改 level=disable:
   - 本机 manifest 记录的 deploy_path -> 直接删(只删 manifest 记录的; 用户手放/内置不碰)
   - 其它机器的记录 -> pending_deletion=true, 那台机器下次 sync 时执行删除并清记录
2. skill 从 manifest 消失后, canonical 仍保留(disable 时), git 里可恢复

设计:纯 json 标准库,无第三方依赖;原子写(tmp + os.replace)防写坏。
"""

from __future__ import annotations

import json
import os
import shutil
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

__all__ = ["DeployRecord", "DeploymentsManifest", "MANIFEST_VERSION"]

MANIFEST_VERSION = 1


def _now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


@dataclass
class DeployRecord:
    skill: str
    machine: str
    agent: str
    deploy_path: str            # 绝对路径(emitter 落盘时展开)
    deployed_at: str = ""       # ISO8601 UTC
    method: str = "cp"          # cp | ln | skipped 不入 manifest(skipped 没部署)
    ir_hash: str = ""           # body sha256(跨机零损耗验证)
    note: str = ""
    pending_deletion: bool = False

    def key(self) -> tuple[str, str, str]:
        """(skill, machine, agent) 唯一键。"""
        return (self.skill, self.machine, self.agent)


@dataclass
class DeploymentsManifest:
    records: list[DeployRecord] = field(default_factory=list)
    path: Optional[Path] = None   # 落盘位置;None = 内存态(测试用)

    # --- load / save ---

    @classmethod
    def load(cls, path: Path) -> "DeploymentsManifest":
        p = Path(path)
        if not p.exists():
            return cls(records=[], path=p)
        with open(p, encoding="utf-8") as fh:
            d = json.load(fh)
        if d.get("version") != MANIFEST_VERSION:
            raise ValueError(f"manifest version {d.get('version')!r} != supported {MANIFEST_VERSION}: {p}")
        records = [DeployRecord(**r) for r in d.get("records", [])]
        return cls(records=records, path=p)

    def save(self, path: Optional[Path] = None) -> None:
        """原子写(tmp + os.replace);不锁,单人编辑流足够。"""
        p = Path(path or self.path)
        if p is None:
            raise ValueError("no manifest path given")
        p.parent.mkdir(parents=True, exist_ok=True)
        d = {"version": MANIFEST_VERSION, "records": [asdict(r) for r in self.records]}
        tmp = p.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(d, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
        os.replace(tmp, p)

    # --- 查询 ---

    def find(self, skill: str, machine: Optional[str] = None, agent: Optional[str] = None) -> list[DeployRecord]:
        out = []
        for r in self.records:
            if r.skill != skill:
                continue
            if machine is not None and r.machine != machine:
                continue
            if agent is not None and r.agent != agent:
                continue
            out.append(r)
        return out

    def skills(self) -> list[str]:
        seen, out = set(), []
        for r in self.records:
            if r.skill not in seen:
                seen.add(r.skill)
                out.append(r.skill)
        return out

    # --- 部署记录生命周期 ---

    def upsert(self, rec: DeployRecord) -> None:
        """按 (skill, machine, agent) 替换或追加。"""
        for i, r in enumerate(self.records):
            if r.key() == rec.key():
                self.records[i] = rec
                return
        if not rec.deployed_at:
            rec.deployed_at = _now_iso()
        self.records.append(rec)

    def remove_record(self, skill: str, machine: str, agent: str) -> Optional[DeployRecord]:
        """删记录(不删盘上文件);返回被删的 record 或 None。"""
        for i, r in enumerate(self.records):
            if r.key() == (skill, machine, agent):
                return self.records.pop(i)
        return None

    # --- 删除链 ---

    def mark_pending_deletion(self, skill: str, except_machine: Optional[str] = None) -> int:
        """rm/disable 时:其它机器的记录标 pending_deletion=true。

        本机(except_machine)的记录由 delete_local 直接处理,不标 pending。
        返回标记条数。
        """
        n = 0
        for r in self.records:
            if r.skill == skill and except_machine is not None and r.machine != except_machine:
                if not r.pending_deletion:
                    r.pending_deletion = True
                    n += 1
        return n

    def delete_local(self, skill: str, machine: str, dry_run: bool = False) -> list[str]:
        """删除链·本机段:删盘上 deploy_path + 清 manifest 记录。

        只删 manifest 记录的路径(用户手放/内置 skill 从不入库,天然不碰)。
        ln 记录取 deploy_path 父目录(软链整个 skill dir);cp 记录取 deploy_path 父目录
        (SKILL.md 在 <dir>/SKILL.md, 删整个 skill 目录)。
        dry_run=True 只报告不动盘。
        返回删除的动作描述列表。
        """
        actions: list[str] = []
        recs = self.find(skill, machine=machine)
        for r in recs:
            target = Path(r.deploy_path)
            # deploy_path 指向 <skill_dir>/SKILL.md(cp 类)或软链的 skill dir(ln 类);
            # 统一删 skill 目录本身:SKILL.md 的父目录 / 软链本身
            if target.name == "SKILL.md":
                target_dir = target.parent
            else:
                target_dir = target
            if dry_run:
                actions.append(f"WOULD DELETE {target_dir} ({r.agent}, {r.method})")
                continue  # dry_run 不动盘也不清记录
            if target_dir.is_symlink():
                target_dir.unlink()
                actions.append(f"unlinked {target_dir} ({r.agent}, {r.method})")
            elif target_dir.exists():
                shutil.rmtree(target_dir)
                actions.append(f"deleted {target_dir} ({r.agent}, {r.method})")
            else:
                actions.append(f"already gone {target_dir} ({r.agent}, {r.method})")
            self.remove_record(skill, machine, r.agent)
        return actions

    def process_pending_deletions(self, machine: str, dry_run: bool = False) -> list[str]:
        """删除链·跨机段:本机 sync 时执行别的机器标记来的 pending_deletion。

        删本机 deploy_path + 清记录。返回动作描述列表。
        """
        actions: list[str] = []
        for r in list(self.records):
            if r.pending_deletion and r.machine == machine:
                target = Path(r.deploy_path)
                target_dir = target.parent if target.name == "SKILL.md" else target
                if dry_run:
                    actions.append(f"WOULD DELETE(pending) {target_dir} ({r.agent})")
                    continue  # dry_run 不动盘也不清记录
                if target_dir.is_symlink():
                    target_dir.unlink()
                    actions.append(f"unlinked(pending) {target_dir} ({r.agent})")
                elif target_dir.exists():
                    shutil.rmtree(target_dir)
                    actions.append(f"deleted(pending) {target_dir} ({r.agent})")
                else:
                    actions.append(f"already gone(pending) {target_dir} ({r.agent})")
                self.remove_record(r.skill, r.machine, r.agent)
        return actions

    # --- 一致性检查(doctor 用) ---

    def check_consistency(self) -> list[str]:
        """盘上文件 vs manifest 记录对账;返回差异描述。"""
        issues = []
        seen_keys = set()
        for r in self.records:
            if r.key() in seen_keys:
                issues.append(f"duplicate record {r.key()}")
            seen_keys.add(r.key())
            target = Path(r.deploy_path)
            if not r.pending_deletion and not target.exists():
                issues.append(f"recorded but missing on disk: {r.deploy_path} ({r.skill}@{r.machine}/{r.agent})")
        return issues

File: internal/manifest/manifest.go
// Package manifest — manifests/deployments.json 读写 + 删除链。
//
// 移植合同(对应 Python manifest.py):
//   - schema v1: {version, records:[DeployRecord...]}
//   - JSON 字节兼容: 字段顺序 = DeployRecord 声明序, indent=2, 非ASCII不转义
//     (ensure_ascii=False), 尾 "\n" — 否则跨机 git diff 全是噪音
//   - 原子写(tmp + rename)防写坏
//
// 删除链(决策 5/6):
//  1. rm <name> / 改 level=disable:
//     - 本机记录的 deploy_path 直接删(只删 manifest 记录的; 用户手放/内置不碰)
//     - 其它机器记录标 pending_deletion=true, 那台机器下次 sync 执行删除
//  2. skill 从 manifest 消失后 canonical 仍保留(disable 时), git 里可恢复
package manifest

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// ManifestVersion — schema 版本。
const ManifestVersion = 1

func nowISO() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05Z")
}

// DeployRecord — 单条部署记录。
// 字段顺序即 JSON 输出序(与 Python dataclass 声明序逐字段对齐 — 字节兼容)。
type DeployRecord struct {
	Skill          string `json:"skill"`
	Machine        string `json:"machine"`
	Agent          string `json:"agent"`
	DeployPath     string `json:"deploy_path"`     // 绝对路径(emitter 落盘时展开)
	DeployedAt     string `json:"deployed_at"`     // ISO8601 UTC
	Method         string `json:"method"`           // cp | ln(skipped 不入 manifest)
	IrHash         string `json:"ir_hash"`          // body sha256(跨机零损耗验证)
	Note           string `json:"note"`
	PendingDeletion bool   `json:"pending_deletion"` // 跨机删除标记
}

// Key — (skill, machine, agent) 唯一键。
type Key struct {
	Skill, Machine, Agent string
}

func (r *DeployRecord) Key() Key {
	return Key{r.Skill, r.Machine, r.Agent}
}

// DeploymentsManifest — manifest 全表。
type DeploymentsManifest struct {
	Records []DeployRecord
	Path    string // 落盘位置; 空 = 内存态(测试用)
}

// Load — 读 manifest; 不存在返回空表。
func Load(path string) (*DeploymentsManifest, error) {
	m := &DeploymentsManifest{Records: nil, Path: path}
	raw, err := os.ReadFile(path)
	if os.IsNotExist(err) {
		return m, nil
	}
	if err != nil {
		return nil, err
	}
	var doc struct {
		Version int            `json:"version"`
		Records []DeployRecord `json:"records"`
	}
	if err := json.Unmarshal(raw, &doc); err != nil {
		return nil, fmt.Errorf("manifest 解析失败 %s: %w", path, err)
	}
	if doc.Version != ManifestVersion {
		return nil, fmt.Errorf("manifest version %d != supported %d: %s", doc.Version, ManifestVersion, path)
	}
	m.Records = doc.Records
	return m, nil
}

// Save — 原子写(tmp + rename)。
// 字节合同: 与 Python json.dumps(ensure_ascii=False, indent=2) + "\n" 全等。
func (m *DeploymentsManifest) Save(path string) error {
	p := path
	if p == "" {
		p = m.Path
	}
	if p == "" {
		return fmt.Errorf("no manifest path given")
	}
	if err := os.MkdirAll(filepath.Dir(p), 0o755); err != nil {
		return err
	}
	doc := struct {
		Version int            `json:"version"`
		Records []DeployRecord `json:"records"`
	}{ManifestVersion, m.Records}
	if doc.Records == nil {
		doc.Records = []DeployRecord{}
	}
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false) // ensure_ascii=False 等价(非ASCII直出 + URL 不转义)
	enc.SetIndent("", "  ")  // indent=2
	if err := enc.Encode(doc); err != nil {
		return err
	}
	// Encoder 自带尾 "\n", 与 Python json.dumps(...) + "\n" 单尾换行等值, 不再补
	out := buf.Bytes()
	tmp := strings.TrimSuffix(p, ".json") + ".json.tmp"
	if err := os.WriteFile(tmp, out, 0o644); err != nil {
		return err
	}
	return os.Rename(tmp, p)
}

// Find — 按 skill(可选 machine/agent 过滤)查记录。
func (m *DeploymentsManifest) Find(skill string, machine, agent string) []DeployRecord {
	var out []DeployRecord
	for _, r := range m.Records {
		if r.Skill != skill {
			continue
		}
		if machine != "" && r.Machine != machine {
			continue
		}
		if agent != "" && r.Agent != agent {
			continue
		}
		out = append(out, r)
	}
	return out
}

// Skills — manifest 里的 skill 名(首次出现序)。
func (m *DeploymentsManifest) Skills() []string {
	seen := map[string]bool{}
	var out []string
	for _, r := range m.Records {
		if !seen[r.Skill] {
			seen[r.Skill] = true
			out = append(out, r.Skill)
		}
	}
	return out
}

// Upsert — 按 (skill, machine, agent) 替换或追加; 新记录补 deployed_at。
func (m *DeploymentsManifest) Upsert(rec DeployRecord) {
	for i, r := range m.Records {
		if r.Key() == rec.Key() {
			m.Records[i] = rec
			return
		}
	}
	if rec.DeployedAt == "" {
		rec.DeployedAt = nowISO()
	}
	m.Records = append(m.Records, rec)
}

// RemoveRecord — 删记录(不删盘上文件); 返回是否删到。
func (m *DeploymentsManifest) RemoveRecord(skill, machine, agent string) bool {
	for i, r := range m.Records {
		if r.Key() == (Key{skill, machine, agent}) {
			m.Records = append(m.Records[:i], m.Records[i+1:]...)
			return true
		}
	}
	return false
}

// MarkPendingDeletion — rm/disable 时: 其它机器的记录标 pending_deletion=true。
// 本机(exceptMachine)记录由 DeleteLocal 直接处理, 不标 pending。返回标记条数。
func (m *DeploymentsManifest) MarkPendingDeletion(skill, exceptMachine string) int {
	n := 0
	for i, r := range m.Records {
		if r.Skill == skill && exceptMachine != "" && r.Machine != exceptMachine {
			if !r.PendingDeletion {
				m.Records[i].PendingDeletion = true
				n++
			}
		}
	}
	return n
}

// DeleteLocal — 删除链·本机段: 删盘上 deploy_path + 清 manifest 记录。
//
// 只删 manifest 记录的路径(用户手放/内置 skill 从不入库, 天然不碰)。
// cp 记录 deploy_path 指 <dir>/SKILL.md → 删 skill 目录; ln 记录指软链目录本身。
// dryRun 只报告不动盘不清记录。返回删除的动作描述列表。
func (m *DeploymentsManifest) DeleteLocal(skill, machine, agent string, dryRun bool) []string {
	var actions []string
	recs := append([]DeployRecord{}, m.Find(skill, machine, agent)...)
	for _, r := range recs {
		targetDir := r.DeployPath
		if filepath.Base(targetDir) == "SKILL.md" {
			targetDir = filepath.Dir(targetDir)
		}
		if dryRun {
			actions = append(actions, fmt.Sprintf("WOULD DELETE %s (%s, %s)", targetDir, r.Agent, r.Method))
			continue
		}
		switch removeSkillDir(targetDir) {
		case removedSymlink:
			actions = append(actions, fmt.Sprintf("unlinked %s (%s, %s)", targetDir, r.Agent, r.Method))
		case removedDir:
			actions = append(actions, fmt.Sprintf("deleted %s (%s, %s)", targetDir, r.Agent, r.Method))
		case alreadyGone:
			actions = append(actions, fmt.Sprintf("already gone %s (%s, %s)", targetDir, r.Agent, r.Method))
		}
		m.RemoveRecord(skill, machine, r.Agent)
	}
	return actions
}

// ProcessPendingDeletions — 删除链·跨机段: 本机 sync 时执行别的机器标来的
// pending_deletion(删本机 deploy_path + 清记录)。返回动作描述列表。
func (m *DeploymentsManifest) ProcessPendingDeletions(machine string, dryRun bool) []string {
	var actions []string
	recs := append([]DeployRecord{}, m.Records...)
	for _, r := range recs {
		if !r.PendingDeletion || r.Machine != machine {
			continue
		}
		targetDir := r.DeployPath
		if filepath.Base(targetDir) == "SKILL.md" {
			targetDir = filepath.Dir(targetDir)
		}
		if dryRun {
			actions = append(actions, fmt.Sprintf("WOULD DELETE(pending) %s (%s)", targetDir, r.Agent))
			continue
		}
		switch removeSkillDir(targetDir) {
		case removedSymlink:
			actions = append(actions, fmt.Sprintf("unlinked(pending) %s (%s)", targetDir, r.Agent))
		case removedDir:
			actions = append(actions, fmt.Sprintf("deleted(pending) %s (%s)", targetDir, r.Agent))
		case alreadyGone:
			actions = append(actions, fmt.Sprintf("already gone(pending) %s (%s)", targetDir, r.Agent))
		}
		m.RemoveRecord(r.Skill, r.Machine, r.Agent)
	}
	return actions
}

type removeOutcome int

const (
	alreadyGone removeOutcome = iota
	removedSymlink
	removedDir
)

// removeSkillDir — 删软链或 skill 目录(与 Python is_symlink→unlink / rmtree 对齐)。
func removeSkillDir(path string) removeOutcome {
	info, err := os.Lstat(path)
	if err != nil {
		if os.IsNotExist(err) {
			return alreadyGone
		}
		return alreadyGone
	}
	if info.Mode()&os.ModeSymlink != 0 {
		if os.Remove(path) == nil {
			return removedSymlink
		}
		return alreadyGone
	}
	if info.IsDir() {
		if os.RemoveAll(path) == nil {
			return removedDir
		}
		return alreadyGone
	}
	// deploy_path 指向普通文件(异常形态): 删文件当目录已删
	if os.Remove(path) == nil {
		return removedDir
	}
	return alreadyGone
}

// CheckConsistency — 盘上文件 vs manifest 记录对账(doctor 用); 返回差异描述。
func (m *DeploymentsManifest) CheckConsistency() []string {
	var issues []string
	seenKeys := map[Key]bool{}
	for _, r := range m.Records {
		k := r.Key()
		if seenKeys[k] {
			issues = append(issues, fmt.Sprintf("duplicate record %s/%s/%s", k.Skill, k.Machine, k.Agent))
		}
		seenKeys[k] = true
		if !r.PendingDeletion {
			if _, err := os.Stat(r.DeployPath); err != nil {
				issues = append(issues, fmt.Sprintf("recorded but missing on disk: %s (%s@%s/%s)",
					r.DeployPath, r.Skill, r.Machine, r.Agent))
			}
		}
	}
	return issues
}